crates/sui-core/src/unit_tests/authority_tests.rs

```diff
@@ -4909,3 +4965,129 @@ fn test_choose_next_system_packages() {
         )
     );
 }
+
+#[tokio::test]
+async fn test_gas_smashing() {
+    // run a create move object transaction with a given set o gas coins and a budget
+    async fn create_obj(
+        sender: SuiAddress,
+        sender_key: AccountKeyPair,
+        gas_coins: Vec<Object>,
+        gas_budget: u64,
+    ) -> (Arc<AuthorityState>, TransactionEffects) {
+        let object_ids: Vec<_> = gas_coins.iter().map(|obj| obj.id()).collect();
+        let (authority_state, pkg_ref) = init_state_with_objects_and_object_basics(gas_coins).await;
+        let effects = create_move_object_with_gas_coins(
+            &pkg_ref.0,
+            &authority_state,
+            &object_ids,
+            gas_budget,
+            &sender,
+            &sender_key,
+        )
+        .await
+        .unwrap();
+        (authority_state, effects)
+    }
+
+    // make a `coin_num` coins distributing `gas_amount` across them
+    fn make_gas_coins(owner: SuiAddress, gas_amount: u64, coin_num: u64) -> Vec<Object> {
+        let mut objects = vec![];
+        let coin_balance = gas_amount / coin_num;
+        for _ in 1..coin_num {
+            let gas_object_id = ObjectID::random();
+            objects.push(Object::with_id_owner_gas_for_testing(
+                gas_object_id,
+                owner,
+                coin_balance,
+            ));
+        }
+        // in case integer division dropped something, make a coin with whatever is left
+        let amount_left = gas_amount - (coin_balance * (coin_num - 1));
+        let gas_object_id = ObjectID::random();
+        objects.push(Object::with_id_owner_gas_for_testing(
+            gas_object_id,
+            owner,
+            amount_left,
+        ));
+        objects
+    }
+
+    // run an object creation transaction with the given amount of gas and coins
+    async fn run_and_check_ok(reference_gas_used: u64, coin_num: u64, budget: u64) -> u64 {
+        let (sender, sender_key): (_, AccountKeyPair) = get_key_pair();
+        let gas_coins = make_gas_coins(sender, reference_gas_used, coin_num);
+        let gas_coin_ids: Vec<_> = gas_coins.iter().map(|obj| obj.id()).collect();
+        let (state, effects) = create_obj(sender, sender_key, gas_coins, budget).await;
+        // transaction is good
+        assert!(effects.status.is_ok());
```

Comment:
Is the only difference between these two functions this one assert?

Reply:
in my classic style I do not look for commonality in tests, but for differences, and I may fork more in coding. I'll check, thanks